#include <iostream>
#include <string>
#include <boost/asio.hpp>
#include <windows.h>
#include <thread>
#include <chrono>
#include <vector>
#include <random>
#include <conio.h> // _kbhit(), _getch() ì‚¬ìš©ì„ ìœ„í•´ ì¶”ê°€

#include "protocol.pb.h"

#pragma pack(push, 1)
struct PacketHeader {
    uint16_t size;
    uint16_t id;
};
#pragma pack(pop)

using boost::asio::ip::tcp;

// =================================================
// ìœˆë„ìš° ì½˜ì†” ë¬¸ìì—´(CP949)ì„ UTF-8ë¡œ ë³€í™˜í•˜ëŠ” í•¨ìˆ˜
// =================================================
std::string AnsiToUtf8(const std::string& ansiStr) {
    if (ansiStr.empty()) return "";
    int wLen = MultiByteToWideChar(CP_ACP, 0, ansiStr.c_str(), -1, NULL, 0);
    std::wstring wStr(wLen, 0);
    MultiByteToWideChar(CP_ACP, 0, ansiStr.c_str(), -1, &wStr[0], wLen);
    int uLen = WideCharToMultiByte(CP_UTF8, 0, wStr.c_str(), -1, NULL, 0, NULL, NULL);
    std::string utf8Str(uLen - 1, 0);
    WideCharToMultiByte(CP_UTF8, 0, wStr.c_str(), -1, &utf8Str[0], uLen, NULL, NULL);
    return utf8Str;
}

std::string GenerateRandomID(int length) {
    const std::string CHARS = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    std::random_device rd;
    std::mt19937 generator(rd());
    std::uniform_int_distribution<> dist(0, CHARS.size() - 1);
    std::string random_string;
    for (int i = 0; i < length; ++i) random_string += CHARS[dist(generator)];
    return random_string;
}

// íŒ¨í‚· ì „ì†¡ í—¬í¼ í•¨ìˆ˜
void SendPacket(tcp::socket& socket, uint16_t pktId, const google::protobuf::Message& msg) {
    std::string payload;
    msg.SerializeToString(&payload);
    PacketHeader header;
    header.size = static_cast<uint16_t>(sizeof(PacketHeader) + payload.size());
    header.id = pktId;

    std::vector<char> send_buffer(header.size);
    memcpy(send_buffer.data(), &header, sizeof(PacketHeader));
    memcpy(send_buffer.data() + sizeof(PacketHeader), payload.data(), payload.size());
    boost::asio::write(socket, boost::asio::buffer(send_buffer));
}

int main() {
    SetConsoleOutputCP(CP_UTF8);
    std::string my_id = GenerateRandomID(6);
    std::string session_token = "";
    std::string gateway_ip = "";
    int gateway_port = 0;

    std::cout << "[DummyClient] Start.. Created by Jeong Shin Young\n";

    try {
        boost::asio::io_context io_context;
        tcp::resolver resolver(io_context);
        tcp::socket socket(io_context);

        // ==========================================
        // LoginServer ì ‘ì† ë° ì›”ë“œ ì„ íƒ ê³¼ì •
        // ==========================================
        std::cout << "[DummyClient] LoginServer(7777) ì—°ê²° ì¤‘...\n";
        boost::asio::connect(socket, resolver.resolve("127.0.0.1", "7777"));

        Protocol::LoginReq login_req;
        login_req.set_id(my_id);
        login_req.set_password("1234");
        SendPacket(socket, Protocol::PKT_CLIENT_LOGIN_LOGIN_REQ, login_req);

        PacketHeader res_header;
        boost::asio::read(socket, boost::asio::buffer(&res_header, sizeof(PacketHeader)));
        std::vector<char> res_payload(res_header.size - sizeof(PacketHeader));
        if (!res_payload.empty()) boost::asio::read(socket, boost::asio::buffer(res_payload.data(), res_payload.size()));

        if (res_header.id == Protocol::PKT_LOGIN_CLIENT_LOGIN_RES) {
            std::cout << "[DummyClient] ê³„ì • ë¡œê·¸ì¸ ì„±ê³µ! ì›”ë“œ(1) ì„ íƒ ìš”ì²­ ì¤‘...\n";
            Protocol::WorldSelectReq ws_req;
            ws_req.set_world_id(1);
            SendPacket(socket, Protocol::PKT_CLIENT_LOGIN_WORLD_SELECT_REQ, ws_req);

            boost::asio::read(socket, boost::asio::buffer(&res_header, sizeof(PacketHeader)));
            res_payload.resize(res_header.size - sizeof(PacketHeader));
            if (!res_payload.empty()) boost::asio::read(socket, boost::asio::buffer(res_payload.data(), res_payload.size()));

            Protocol::WorldSelectRes w_res;
            if (w_res.ParseFromArray(res_payload.data(), res_payload.size()) && w_res.success()) {
                session_token = w_res.session_token();
                gateway_ip = w_res.gateway_ip();
                gateway_port = w_res.gateway_port();
                std::cout << "[DummyClient] ğŸ‰ ì›”ë“œ ì…ì¥ ìŠ¹ì¸ ì™„ë£Œ! í† í° ë°œê¸‰ë¨.\n";
            }
            else return 0;
        }

        // ==========================================
        // LoginServer ì—°ê²° ì¢…ë£Œ ë° GatewayServer ì—°ê²°
        // ==========================================
        socket.close();
        std::cout << "--------------------------------------\n";
        std::cout << "[DummyClient] GatewayServer(" << gateway_port << ") ë¡œ ê²Œì„ ì§„ì…ì„ ì‹œë„í•©ë‹ˆë‹¤...\n";

        socket.open(tcp::v4());
        boost::asio::connect(socket, resolver.resolve(gateway_ip, std::to_string(gateway_port)));

        Protocol::GatewayConnectReq gw_req;
        gw_req.set_account_id(my_id);
        gw_req.set_session_token(session_token);
        SendPacket(socket, Protocol::PKT_CLIENT_GATEWAY_CONNECT_REQ, gw_req);

        boost::asio::read(socket, boost::asio::buffer(&res_header, sizeof(PacketHeader)));
        res_payload.resize(res_header.size - sizeof(PacketHeader));
        if (!res_payload.empty()) boost::asio::read(socket, boost::asio::buffer(res_payload.data(), res_payload.size()));

        if (res_header.id == Protocol::PKT_GATEWAY_CLIENT_CONNECT_RES) {
            std::cout << "======================================\n";
            std::cout << "       ğŸ•¹ï¸ ì¸ê²Œì„ ì„¸ê³„ì— ì§„ì…í–ˆìŠµë‹ˆë‹¤!       \n";
            std::cout << "======================================\n";
        }

        // ================================================
        // ì¸ê²Œì„ í†µì‹ : ìˆ˜ì‹  ìŠ¤ë ˆë“œì™€ ë©”ì¸ ë£¨í”„ (ëª¨ë“œ ì „í™˜)
        // ================================================

        // [ì„œë²„ íŒ¨í‚· ìˆ˜ì‹  ì „ìš© ë°±ê·¸ë¼ìš´ë“œ ìŠ¤ë ˆë“œ]
        std::thread recv_thread([&socket, my_id]() {
            try {
                while (true) {
                    PacketHeader h;
                    boost::asio::read(socket, boost::asio::buffer(&h, sizeof(PacketHeader)));
                    std::vector<char> p(h.size - sizeof(PacketHeader));
                    if (!p.empty()) boost::asio::read(socket, boost::asio::buffer(p.data(), p.size()));

                    if (h.id == Protocol::PKT_GATEWAY_CLIENT_CHAT_RES) {
                        Protocol::ChatRes chat_res;
                        if (chat_res.ParseFromArray(p.data(), p.size())) {
                            std::cout << "\n[ì±„íŒ…] " << chat_res.account_id() << " : " << chat_res.msg() << "\n";
                        }
                    }
                    else if (h.id == Protocol::PKT_GATEWAY_CLIENT_MOVE_RES) {
                        Protocol::MoveRes move_res;
                        if (move_res.ParseFromArray(p.data(), p.size())) {
                            // ë‚´ê°€ ì›€ì§ì¸ ê²°ê³¼ëŠ” í™”ë©´ ë„ë°° ë°©ì§€ë¥¼ ìœ„í•´ ìˆ¨ê¸°ê³ , ë‹¤ë¥¸ ìœ ì €ì˜ ì´ë™ë§Œ ì¶œë ¥í•©ë‹ˆë‹¤.
                            //if (move_res.account_id() != my_id) {
                            //    std::cout << "\n[ì´ë™] ìœ ì €(" << move_res.account_id() << ") -> X:" << move_res.x() << " Y:" << move_res.y() << "\n";
                            //}
                        }
                    }
                }
            }
            catch (...) { std::cout << "\n[ì„œë²„ ì—°ê²° ì¢…ë£Œ]\n"; }
            });
        recv_thread.detach();


        // í•˜ë‚˜ì˜ ì„¸ë ¨ëœ ë…¼ë¸”ë¡œí‚¹ í‚¤ë³´ë“œ ì œì–´ ë£¨í”„ë¡œ í†µí•©

        float my_x = 0.0f, my_y = 0.0f;
        std::cout << "\n [ì•¡ì…˜ ëª¨ë“œ] ë°©í–¥í‚¤: ì´ë™ / Enter: ì±„íŒ… / ESC: ì¢…ë£Œ\n";
        std::cout << "--------------------------------------\n";

        // [ì‚¬ìš©ì ì…ë ¥ ì „ìš© ë©”ì¸ ë£¨í”„ (ì•¡ì…˜/ì±„íŒ… ëª¨ë“œ ì œì–´)]
        while (true) {
            // í‚¤ë³´ë“œ ì…ë ¥ì´ ìˆì„ ë•Œë§Œ ë°˜ì‘ (ë¸”ë¡œí‚¹ ë˜ì§€ ì•ŠìŒ)
            if (_kbhit()) {
                int key = _getch();

                // 1. ë°©í–¥í‚¤ ì…ë ¥ ê°ì§€ (íŠ¹ìˆ˜í‚¤ëŠ” 224ê°€ ë¨¼ì € ë“¤ì–´ì˜µë‹ˆë‹¤)
                if (key == 224) {
                    key = _getch();
                    bool moved = false;
                    switch (key) {
                    case 72: my_y += 1.0f; moved = true; break; // UP
                    case 80: my_y -= 1.0f; moved = true; break; // DOWN
                    case 75: my_x -= 1.0f; moved = true; break; // LEFT
                    case 77: my_x += 1.0f; moved = true; break; // RIGHT
                    }

                    if (moved) {
                        Protocol::MoveReq move_req;
                        move_req.set_x(my_x);
                        move_req.set_y(my_y);
                        move_req.set_z(0.0f);
                        move_req.set_yaw(0.0f);
                        SendPacket(socket, Protocol::PKT_CLIENT_GATEWAY_MOVE_REQ, move_req);

                        // \r ì„ ì‚¬ìš©í•´ ê°™ì€ ì¤„ì—ì„œ ë‚´ ì¢Œí‘œë§Œ ì‹¤ì‹œê°„ ê°±ì‹ í•©ë‹ˆë‹¤.
                        std::cout << "[ë‚´ ìœ„ì¹˜] X:" << my_x << " Y:" << my_y << "          \r";
                    }
                }
                // 2. Enter í‚¤ (13) ëˆ„ë¦„ -> [ì±„íŒ… ëª¨ë“œ] ì§„ì…
                else if (key == 13) {
                    std::cout << "\n[ì±„íŒ… ëª¨ë“œ] ì…ë ¥> ";
                    std::string input;
                    // ì—¬ê¸°ì„œë§Œ ì¼ì‹œì ìœ¼ë¡œ std::getlineì´ ì‹¤í–‰ë˜ì–´ íƒ€ì´í•‘ì„ ë°›ìŠµë‹ˆë‹¤.
                    std::getline(std::cin, input);

                    if (!input.empty()) {
                        Protocol::ChatReq chat_req;
                        chat_req.set_msg(AnsiToUtf8(input));
                        SendPacket(socket, Protocol::PKT_CLIENT_GATEWAY_CHAT_REQ, chat_req);
                    }
                    std::cout << "[ì•¡ì…˜ ëª¨ë“œ] ë°©í–¥í‚¤: ì´ë™ / Enter: ì±„íŒ… / ESC: ì¢…ë£Œ\n";
                }
                // 3. ESC í‚¤ (27) ëˆ„ë¦„ -> í”„ë¡œê·¸ë¨ ì¢…ë£Œ
                else if (key == 27) {
                    std::cout << "\n[DummyClient] ì ‘ì†ì„ ì¢…ë£Œí•©ë‹ˆë‹¤.\n";
                    break;
                }
            }

            // CPU 100% ì ìœ  ë°©ì§€ë¥¼ ìœ„í•œ ì§§ì€ íœ´ì‹
            std::this_thread::sleep_for(std::chrono::milliseconds(10));
        }

        socket.close();
    }
    catch (std::exception& e) { std::cerr << "[Error] " << e.what() << "\n"; }
    return 0;
}